File: Day275/constrainedSubSequenceSum.go
package main

import (
	"fmt"
	"math"
)
func constrainedSubsetSum(nums []int, k int) int {
    queue := make([]int, 0)
    for i:= range nums {
        if len(queue) != 0 {
            nums[i] += nums[queue[0]]
        }
        for len(queue) != 0 && (i-queue[0] >= k || nums[i] >= nums[queue[len(queue)-1]]) {
            if nums[i] >= nums[queue[len(queue)-1]] {
                queue = queue[:len(queue)-1]
            } else {
                queue = queue[1:]
            }
        }
        if nums[i] > 0 {
            queue = append(queue, i)
        }
    }
    max := nums[0]
    for i:=1; i<len(nums); i++ {
        max = int(math.Max(float64(max), float64(nums[i]) ))
    }
    return max
}
func main() {
	k2 := 2
	k1 := 2
	k3 := 3
	nums1 := []int{10,2, -10, 5, 20}
	nums2 := []int{-1,-2, -2}
	nums3 := []int{10,-2, -10, -5,20}
	fmt.Println(constrainedSubsetSum(nums1, k1));
	fmt.Println(constrainedSubsetSum(nums2, k2));
	fmt.Println(constrainedSubsetSum(nums3, k3));
}

File: Day275/constrainedSubSequenceSum.java
package Day275;

import java.util.Arrays;
import java.util.PriorityQueue;

public  class constrainedSubSequenceSum {
    public static int constrainedSubsetSum(int[] nums, int k) {
        int len = nums.length;
        int arr[] = new int[len];
        int answer = nums[0];
        PriorityQueue<Integer> priorityQueue = new PriorityQueue<>((a,b)->arr[b]-arr[a]);
        Arrays.fill(arr, Integer.MIN_VALUE);
        arr[0] = nums[0];
        priorityQueue.offer(0);
        for (int i=1; i<len; i++) {
            int j = Math.max(i-k,0);
            while (!priorityQueue.isEmpty() &&priorityQueue.peek()<j) {
                priorityQueue.poll();
            }
            int ind = priorityQueue.peek();
            arr[i] = Math.max(arr[ind] + nums[i], nums[i]);
            answer = Math.max(answer, arr[i]);
            priorityQueue.offer(i);
        }
        return answer;
    }
    public static void main(String[] args) {
        int k1 = 2;
        int k2 = 2;
        int k3 = 3;
        int nums1[] = {10,2, -10, 5, 20};
        int nums2[] = {-1,-2, -2};
        int nums3[] = {10,-2, -10, -5,20};
        System.out.println(constrainedSubsetSum(nums1, k1));
        System.out.println(constrainedSubsetSum(nums2, k2));
        System.out.println(constrainedSubsetSum(nums3, k3));

    }
}